deployment/terraform/batch.tf

```diff
@@ -0,0 +1,75 @@
+#
+# Batch resources
+#
+resource "aws_launch_template" "batch_gpu_container_instance" {
+  block_device_mappings {
+    device_name = "/dev/xvda"
+
+    ebs {
+      volume_size = 64
+      volume_type = "gp2"
+    }
+  }
+}
+
+# Pull the AMI ID for the latest Amazon ECS GPU-optimized AMI
+# https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-optimized_AMI.html#w520aac22c15c31b5
+data "aws_ssm_parameter" batch_gpu_ami_id {
```

Comment:
I am worried that as this AMI ID changes over time, we are going to run into issues when Terraform attempts to destroy and recreate the CE. This worry is increased due to our desire to pursue CD.

If this value changes, does the CE swap happen seamlessly?

Reply:
Agreed, if this value changes, Terraform will attempt to destroy and recreate the CE. I missed that.

There's still an elegant solution. We can define `image_id` within the launch template instead of the compute environment. What do you think?

Reply:
Agreed, if this value changes, Terraform will attempt to destroy and recreate the CE. I missed that.

There's still an elegant solution. We can define `image_id` within the launch template instead of the compute environment. What do you think?

Reply:
You may also find this recent enhancement to the `aws_batch_compute_environment` resource helpful for replacement scenarios. 😄 https://github.com/terraform-providers/terraform-provider-aws/issues/11077#issuecomment-570673120

Reply:
I haven't tested the launch template maneuver yet, but looking at the [CloudFormation docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-batch-computeenvironment-computeresources.html#cfn-batch-computeenvironment-computeresources-launchtemplate) for Batch CE, updates to launch templates require replacement too.

Given that, I'd be inclined to pursue one of the following paths:

1. Update this PR so that a static Batch AMI ID is provided to the Batch CE; open an issue to use `compute_environment_name_prefix` later; merge sooner
2. Test the entire project with the most current version of the AWS provider to ensure there aren't any noticable regressions; wait until [Thursday](https://github.com/terraform-providers/terraform-provider-aws/issues/11077#issuecomment-570673120) for 2.44.0 and update the PR to make use of `compute_environment_name_prefix`; merge later

(Thanks for the long-distance assist, @bflad. 😎)

Reply:
Right, updates to launch templates require replacement of the Batch CE, but the launch template specification uses the default version of the launch template. The thinking was we update the launch template, but keep the pointer in the Batch CE the same.

I tested this and thought it was working, however, it looks like Terraform doesn't set the default version after updating a launch template (terraform-providers/terraform-provider-aws/issues/4655).

Following a suggestion in that issue, I recreated the Batch CE with the version set to `$Latest`. Strangely enough, the Batch CE is still launching instances with version 1 of the launch template.

Given all that, I think we should follow the second path you've outlined.

Reply:
Just a heads up that it looks like the 2.44.0 release has been published.